File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.py
class Solution:
    def maxTwoEvents(self, events: List[List[int]]) -> int:
        answer = 0
        n = len(events)
        events.sort()

        max_value = [0] * n
        max_value[-1] = events[-1][2]

        for i in range(n-2, -1, -1):
            max_value[i] = max(max_value[i+1], events[i][2])

        for i, event in enumerate(events):
            answer = max(answer, event[2])
            left, right = 0, n-1
            idx = -1
            while left <= right:
                mid = (left + right) // 2

                if events[mid][0] > event[1]:
                    idx = mid
                    right = mid - 1
                else:
                    left = mid + 1

            if idx != -1:
                answer = max(answer, event[2] + max_value[idx])

        return answer

File: LeetCode/2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.java
class Solution {
    public int maxTwoEvents(int[][] events) {
        int answer = 0;
        
        Arrays.sort(events, (a,b) -> a[0] - b[0]);
        int n = events.length;
        
        int[] maxValue = new int[n];
        
        maxValue[n-1] = events[n-1][2];
        
        for(int i = n - 2; i >= 0; i--){
            maxValue[i] = Math.max(events[i][2], maxValue[i+1]);
        }
        
        int left, right, mid, idx;
        
        for(int i = 0; i < n; i++){
            //System.out.println(answer);
            answer = Math.max(answer, events[i][2]);
            left = 0;
            right = n-1;
            idx = -1;
            
            while(left <= right){
                mid = (left+right)/2;
                
                if(events[mid][0] > events[i][1]){
                    right = mid -1;
                    idx = mid;
                }
                else{
                    left = mid + 1;
                }
            }
            if(idx != -1){
                answer = Math.max(answer, events[i][2] + maxValue[idx]);
            }
            
        }
        
        return answer;
    }
}